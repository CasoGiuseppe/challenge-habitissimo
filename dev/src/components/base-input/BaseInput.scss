.base-input {
  $baseinput: &;

  @include text--regular;
  @include moveValidationUI;

  display: flex;
  flex-direction: column;
  position: relative;
  
  &__label {
    background: $c-white;
    border: 1px solid rgba($c-gray, .2);
    border-radius: px2rem($base);
    display: flex;
    padding: #{($base * 3) + 3px} #{$base * 2px} #{$base + px};
    position: relative;
    transition: border .4s $a-ease-out;

    &::before,
    &::after {
      content: '';
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
    }

    &::before {
      @include text--light;
      @include text--xsmall;
      content: attr(data-placeholder);
      color: $c-gray;
      transition: transform .4s $a-ease-out;
    }

    #{$baseinput}--isActive &{
      border-color: $c-secondary;

      &::before {
        transform: translateY(-50%) translateY(calc((100% - #{$base + px}) * -1));
      }
    }

    #{$baseinput}--isValid &{
      border-color: $c-valid;
    }

    #{$baseinput}--isInvalid &{
      border-color: $c-warning;
    }
  }

  &__field {
    @include text--xsmall;
    background: none;
    border: none;
    color: $c-dark-gray;
    padding: #{$base - 2px};
    width: 100%;
  }

  &__required-label {
    @include text--xxsmall;
    align-self: flex-end;
    color: $c-gray;
  }

  &__validation-icon {
    background: $c-silver;
    border-radius: 100%;
    height: #{$base * 3px};
    position: absolute;
    top: 50%;
    left: 0;
    margin-top: #{(($base * 3) / 2) * -1px};
    width: #{$base * 3px};
    transform: translateX(-100%) translateX(#{$base * -1px});
    transition: background .4s linear;

    #{$baseinput}--isValid &{
      @include UI-isValid
    }

    #{$baseinput}--isInvalid &{
      @include UI-isInvalid
    }
  }
}
