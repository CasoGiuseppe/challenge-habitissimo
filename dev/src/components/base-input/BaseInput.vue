<template>
  <div class="base-input">
    <label
      :class="[
        'base-input__label',
        onActive ? 'base-input__label--isActive' : null,
        Object.prototype.hasOwnProperty.call(this.isValidate, 'state') && !isValidate.state
          ? 'base-input__label--isInvalid'
          : null,
        Object.prototype.hasOwnProperty.call(this.isValidate, 'state') && isValidate.state
          ? 'base-input__label--isValid'
          : null,
      ]"
      :for="id"
      :data-placeholder="placeholder">
      <input
        class="base-input__field"
        :type="type"
        :id="id"
        :required="isRequired.state"
        @focus="handleActiveState"
        @blur="handleActiveState"
        @input="handleInput"/>
      <!-- UI validation icon -->
      <transition
        mode="out-in"
        name="change-validation">
          <span
            class="base-input__validation-icon"
            v-if="Object.prototype.hasOwnProperty.call(this.isValidate, 'state')"
            :key="isValidate.state">
          </span>
      </transition>
    </label>

    <!-- required label -->
    <span
      class="base-input__required-label"
      v-if="isRequired.state">
        {{isRequired.label}}
    </span>
  </div>
</template>

<script>
export default {
  name: 'BaseInput',
  data() {
    return {
      onFocus: false,
      onActive: false,
    };
  },

  props: {
    type: {
      type: String,
      default: 'text',
    },

    id: {
      type: String,
      default: 'test',
    },

    placeholder: {
      type: String,
      default: 'placeholder',
    },

    isRequired: {
      type: Object,
      default() {
        return {
          state: false,
          label: 'required',
        };
      },
    },

    isValidate: {
      type: Object,
      default() {
        return {};
      },
    },
  },

  methods: {
    handleInput(e) {
      this.$emit('input', e.target.value);
    },

    handleActiveState(e) {
      const { type } = e;

      if (type === 'blur') {
        this.onActive = e.target.value !== '';
      } else {
        this.onActive = true;
        delete this.isValidate.state;
      }
    },
  },
};
</script>

<style lang='scss' scoped src= './BaseInput.scss'></style>
