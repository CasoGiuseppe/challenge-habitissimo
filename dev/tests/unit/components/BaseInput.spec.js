import BaseInput from '@/components/base-input/BaseInput'
import { mount } from '@vue/test-utils'

describe('BaseInput.vue', () => {
  let component

  beforeEach(() => {
    component = mount(BaseInput, {
      propsData:{
        type: 'text',
        id: 'test',
        placeholder: 'placeholder',
        isRequired: {
          state: false,
          label: 'required'
        },
        isValidate: {
          state: true
        }
      }
    })
  })

  it('has correct props values', () => {
    expect(component.props().type).toBe('text')
    expect(component.props().id).toBe('test')
    expect(component.props().placeholder).toBe('placeholder')
    expect(component.props().isRequired).toHaveProperty('label', 'required')
    expect(component.props().isValidate).toHaveProperty('state', true)
  })

  it('has correct $emit response', async () => {
    component.vm.$emit('input', 'inputvalue')

    await component.vm.$nextTick()
    expect(component.emitted().input).toBeTruthy()
    expect(component.emitted().input[0]).toEqual(['inputvalue'])
  })

  it('has correct set data value', () => {
    component.setData({onActive: true})
    expect(component.vm.onActive).toBeTruthy()
  })

  it('has input correct ID', () => {
    const input = component.find('.base-input__field')
    expect(input.attributes().id).toEqual('test')
  })

  it('has the expected html structure', () => {
    expect(component.html()).toMatchSnapshot()
  })
})