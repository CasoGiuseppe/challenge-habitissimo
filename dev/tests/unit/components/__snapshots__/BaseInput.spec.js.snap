// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput.vue has the expected html structure 1`] = `
<div class="base-input base-input--isValid"><label for="test" data-placeholder="placeholder" class="base-input__label"><input id="test" type="text" class="base-input__field">
    <transition-stub name="change-validation" mode="out-in"><span class="base-input__validation-icon"></span></transition-stub>
  </label>
  <!---->
</div>
`;
